documentation/architecture/authorization/ABAC.md

```diff
@@ -0,0 +1,194 @@
+
+Attributes are divided into three categories:
+
+- Subject attributes, belonging to the caller
+- Action attributes, describing the action
+- Resource attributes, belonging to resource being accessed
+
+For example subject can be user in the system, and it has an attribute "email"
+Action can be a type of request, for example "Update"
+Action can also have additional attributes, like "field"
+And resource is a project we're trying to access, which can have an attribute "owners"
+
+Then we can define our access policy to only let project owners update "space" field as:
+
+To perform an action "update", if action attribute "field" is "services", the subject attribute "email" should be a member of a list, which is a resource attribute "owners".
+
+### Policy enforcement and decision points
+
+Policy Enforcement Point (Integration) - a place in the code, which using the current code context
+can create and ABAC request.
+
+Policy Decision Point (Policy Check) - system which can decide whether request is authorized or not.
+
+Policy Information Point (Policy Storage) - a storage which keeps information about policies.
+
+PDP and PIP can be abstracted and implemented as a separate system and even delegated
+to remote nodes.
+
+PEP is an integration point to convert existing request context into ABAC terms.
+
+For example in message flow authorization PEP is an `is_authorized` function.
+It takes a message and a worker state and creates an ABAC request which contains
+subject, action and resource attributes.
+
+### ABAC requests and policy rules
+
+ABAC request is a set of attributes derived from the current request and context by PEP.
+
+ABAC policy is a set of rules applied to the attributes.
+
+In order to make request-policy association easier, both contain ActionID, which
+describes a resource and an action being used.
+This allows for example to use it as a key in the database.
+
+For example ActionID for example above can be written as `Project/Update`
+
+#### Policy rules
+
+Policy rules for Ockam ABAC are defined as s-expressions.
+Each rule resolves into boolean value when "true" means "allow".
+
+##### Bool rule:
+
+- `true`
+- `false`
+
+Mostly useful in logical rules, but can be used as allow-all or deny-all policies
+
+##### Comparison rules:
+
+- equality `(= subject.foo "bar")` `(!= subject.foo "bar")`
+- numeric comparison `(> subject.foo 1)`, `(< subject.foo 3)`
+
+Supports operators `=` `!=` `>` and `<`
+First argument MUST be an Attribute
+Second argument can be an Attribute or a Value
+
+##### Inclustion rule:
+
+Matches that an attribute belongs to a list.
+
+- `(member? subject.key [1 2 3])`
+- `(member? subject.key resource.keys)`
+- `(member? "foo" action.keys)`
+
+First argument can be an Attribute or a Value
+Second argument can be an Attribute or a List
+If second argument is an Attribute, the corresponding Attribute should resolve to a List
+
+##### Logical rules:
+
+Combine other rules
+
+- `(not (in "foo" action.keys))`
```

Comment:
Is there a difference between `in` and `member?` used above?

Reply:
We decided to not use `in`, will update